tendermint/src/time.rs

```diff
@@ -150,11 +144,145 @@ pub trait ParseTimestamp {
 
 #[cfg(test)]
 mod tests {
+    use std::convert::TryInto;
+
     use super::*;
 
-    #[test]
-    fn serde_roundtrip() {
-        const DATES: &[&str] = &[
+    // TODO(shon) Extract arbitrary generators into own library
+
+    use chrono::{DateTime, NaiveDate, TimeZone, Timelike, Utc};
+    use proptest::{prelude::*, sample::select};
+
+    // Any higher, and we're at seconds
+    const MAX_NANO_SECS: u32 = 999_999_999u32;
+
+    // With values larger or smaller then these, chrono produces invalid rfc3339
+    // timestamps. See https://github.com/chronotope/chrono/issues/537
+    fn min_time() -> DateTime<Utc> {
+        Utc.timestamp(-9999999999, 0)
+    }
+
+    fn max_time() -> DateTime<Utc> {
+        Utc.timestamp(99999999999, 0)
+    }
+
+    fn num_days_in_month(year: i32, month: u32) -> u32 {
```

Comment:
What does this method add beyond something trivial like:

```rust
fn num_days_in_month(year: i32, month: u32) -> u32 {
    match month {
        2 => if year % 4 == 0 { 29 } else { 28 },
        1 | 3 | 5 | 7 | 8 | 10 | 12 => 31,
        4 | 6 | 9 | 11 => 30,
    }
}
```

Reply:
It just offloads the calculation of dates to the datetime library we've imported for that very purpose :) 

We can compute it ourselves, but, assuming we are using the Gregorian calendar, it's slightly more involved than checking for divisibility by 4. See https://docs.microsoft.com/en-us/office/troubleshoot/excel/determine-a-leap-year#how-to-determine-whether-a-year-is-a-leap-year 

I don't feel super strongly that we must use chrono here. But as a matter of principle, if I'm using an external library to deal with some tricky domain (and time is tricky), I prefer to consistently delegate to the library wherever possible and not substitute in my own home-rolled algorithms. I find the latter is often a source of subtle bugs, since I usually don't haven't deeply internalized all the aspects of the domain (which is why I'm using the external lib in the first place :)).

Reply:
Makes sense!